### PR TITLE
Fix beatmap updates causing one extra carousel selection

### PR DESCRIPTION
**Pull request by peppy**
This fixes the carousel making one unnecessary selection every time the current selection is updated. Basically, on returning from gameplay this will cause the selection sound to no longer (incorrectly) play, but also likely improves smoothness as it's not randomly selecting a beatmap *somewhere* in the carousel before correcting back to the retained selection.

I'd like to let the code speak for itself here. The only reason I did this is that I was already so deep that it seemed worthwhile. The resultant code does seem like an improvement over what we had, but I'll let others be the judge.

- [x] Depends on https://github.com/ppy/osu/pull/25943

### CODE REVIEW

File: osu.Game/Screens/Select/BeatmapCarousel.cs [hunk 1]
```diff
@@ -1207,6 +1184,43 @@ public override void AddItem(CarouselItem i)
...
+            /// <summary>
+            /// A special method to handle replace operations (general for updating a beatmap).
+            /// Avoids event-driven selection flip-flopping during the remove/add process.
+            /// </summary>
+            /// <param name="oldItem">The beatmap set to be replaced.</param>
+            /// <param name="newItems">All new items to replace the removed beatmap set.</param>
+            /// <returns>All removed items, for any further processing.</returns>
+            public IEnumerable<CarouselBeatmapSet> ReplaceItem(BeatmapSetInfo oldItem, List<CarouselBeatmapSet> newItems)
+            {
+                var previousSelection = (LastSelected as CarouselBeatmapSet)?.Beatmaps
```
Comment (bdach) [inline osu.Game/Screens/Select/BeatmapCarousel.cs:1196]: this is generally unrelated commentary but the fact that this class has both `LastSelected` and `selectedBeatmapSet` that are separately managed and do not delegate to one another is... just great

**Review by bdach — DISMISSED** [review]
code-wise i don't see this being much better than the previous version but it fixes an audio issue so let's see if this holds up

**Comment by bdach** [comment]
i bet you'll love this test failure: https://github.com/ppy/osu/actions/runs/7274690433/job/19821122908?pr=25949#step:5:34

test covers exact code touched, so looks suspicious to say the least.

File: osu.Game/Screens/Select/BeatmapCarousel.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs:1196]: Do you mean the parent class in this case?

**Comment by peppy** [comment]
> i bet you'll love this test failure: https://github.com/ppy/osu/actions/runs/7274690433/job/19821122908?pr=25949#step:5:34

Weird enough, I think I managed to repro this on master at some point during local testing. I'll take a look.

File: osu.Game/Screens/Select/BeatmapCarousel.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Screens/Select/BeatmapCarousel.cs:1196]: yeah. i'm not saying anything has to be done about it but yeah

**Comment by bdach** [comment]
As per https://discord.com/channels/188630481301012481/188630652340404224/1187038966797566072 this fails elsewhere, so merging.

**Review by bdach — APPROVED** [review]

**Pull request merged**